src/peer/handshaking/peer_handshaking_state.rs

```diff
@@ -1,110 +1,188 @@
+use tezos_messages::p2p::encoding::prelude::{AckMessage, MetadataMessage};
 
-use crate::tmp::chunking::ChunkReadBuffer;
-use crypto::crypto_box::PublicKey;
-use tezos_messages::p2p::binary_message::BinaryChunk;
-use tezos_messages::p2p::encoding::ack::NackMotive;
-use tezos_messages::p2p::encoding::prelude::{AckMessage, MetadataMessage, NetworkVersion};
-
-use crate::io_error_kind::IOErrorKind;
-use crate::peer::{PeerCrypto, PeerToken};
-use crate::Port;
+use crate::peer::binary_message::read::peer_binary_message_read_state::PeerBinaryMessageReadState;
+use crate::peer::binary_message::write::peer_binary_message_write_state::PeerBinaryMessageWriteState;
+use crate::peer::chunk::read::peer_chunk_read_state::{PeerChunkReadState, ReadCrypto};
+use crate::peer::chunk::write::peer_chunk_write_state::{PeerChunkWriteState, WriteCrypto};
+use crate::peer::PeerToken;
 
 #[derive(Serialize, Deserialize, Debug, Clone)]
-pub enum MessageWriteState {
-    Idle,
-    Pending { written: usize },
-    Success,
-    // TODO: use custom error instead.
-    Error { error: IOErrorKind },
+pub enum PeerHandshakingError {
+    Chunk(String),
+    Crypto(String),
+    Encoding(String),
+    Decoding(String),
+    Blake2b(String),
 }
 
-impl MessageWriteState {
-    #[inline(always)]
-    pub fn is_idle(&self) -> bool {
-        matches!(self, Self::Idle)
+impl From<CryptoError> for PeerHandshakingError {
+    fn from(error: CryptoError) -> Self {
+        Self::Crypto(error.to_string())
     }
+}
 
-    #[inline(always)]
-    pub fn is_success(&self) -> bool {
-        matches!(self, Self::Success)
+impl From<BinaryWriterError> for PeerHandshakingError {
+    fn from(error: BinaryWriterError) -> Self {
+        Self::Encoding(error.to_string())
     }
 }
 
-#[derive(Serialize, Deserialize, Debug, Clone)]
-pub enum MessageReadState<M> {
-    Idle,
-    Pending { buffer: ChunkReadBuffer },
-    Success { message: M },
-    // TODO: use custom error instead.
-    Error { error: IOErrorKind },
+impl From<BinaryReaderError> for PeerHandshakingError {
+    fn from(error: BinaryReaderError) -> Self {
+        Self::Decoding(error.to_string())
+    }
 }
 
-#[derive(Serialize, Deserialize, Debug, Clone)]
-pub struct ConnectionMessageDataReceived {
-    pub port: Port,
-    pub compatible_version: Option<NetworkVersion>,
-    pub public_key: PublicKey,
-    pub encoded: BinaryChunk,
+impl From<BinaryChunkError> for PeerHandshakingError {
+    fn from(error: BinaryChunkError) -> Self {
+        Self::Chunk(error.to_string())
+    }
+}
+
+impl From<Blake2bError> for PeerHandshakingError {
+    fn from(error: Blake2bError) -> Self {
+        Self::Blake2b(error.to_string())
+    }
 }
 
 #[derive(Serialize, Deserialize, Debug, Clone)]
 pub enum PeerHandshakingStatus {
     Init,
-    ConnectionMessageWrite {
-        /// Encoded `ConnectionMessage` to be sent.
-        conn_msg: BinaryChunk,
-        status: MessageWriteState,
+    ConnectionMessageInitialized {
```

Comment:
Please use `Init` here as well, there are other states like this one and change there also.

Lets try to follow vocabulary: "init", "pending", "success", "error", as close as we can.

Reply:
Well, `Init` is for `Initial`, and `Initialized` might be slightly different. But let's have `Init` for now.